### PR TITLE
Issue 880

### LINKED ISSUE #880: Initialize engine.state before calling run
## 🚀 Feature

Idea is to improve flexibility of Engine and simplify user code by initializing `engine.state` before calling `run`.

What we gain:
```python
trainer = Engine(...)
trainer.state.alpha = 0.0
trainer.state.beta = 1.0
```

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #880  

Description:
- engine.state is created at engine init

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/contrib/handlers/test_lr_finder.py [hunk 1]
```diff
@@ -136,11 +136,14 @@ def test_model_optimizer_reset(lr_finder, to_save, dummy_engine, dataloader):
...
 
-    with lr_finder.attach(dummy_engine, to_save=to_save, diverge_th=float("inf")) as trainer_with_finder:
-        trainer_with_finder.run(dataloader)
+    with pytest.warns(UserWarning, match=r"Run completed without loss diverging"):
+        with lr_finder.attach(dummy_engine, to_save=to_save, diverge_th=float("inf")) as trainer_with_finder:
+            trainer_with_finder.run(dataloader)
 
     assert init_optimizer_sd == optimizer.state_dict()
     assert init_model_sd == model.state_dict()
+    print(init_trainer_sd)
```
Comment (sdesrozis) [inline tests/ignite/contrib/handlers/test_lr_finder.py:145]: To remove

File: ignite/contrib/handlers/lr_finder.py [hunk 2]
```diff
@@ -336,9 +338,10 @@ def attach(
             self._detach(trainer)
             # restore to_save and reset trainer's state
             obj = torch.load(cache_filepath.as_posix())
-            trainer.state = None
+            trainer.load_state_dict(obj["trainer"])
```
Comment (sdesrozis) [inline ignite/contrib/handlers/lr_finder.py:341]: Can it conflict with others loads ?

File: tests/ignite/contrib/handlers/test_lr_finder.py [hunk 3]
```diff
@@ -136,11 +136,14 @@ def test_model_optimizer_reset(lr_finder, to_save, dummy_engine, dataloader):
...
+    print(dummy_engine.state_dict())
```
Comment (sdesrozis) [inline tests/ignite/contrib/handlers/test_lr_finder.py:146]: To remove

**Comment by vfdev-5** [comment]
> I don't understand yet why lr_finder implies to save trainer but it's a detail :)

because when lr finder is run, engine is run too. Thus we need to come back to the initial state.

File: ignite/contrib/handlers/lr_finder.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/contrib/handlers/lr_finder.py:341]: > Can it conflict with others loads ?

Which others loads ? 
I didn't include trainer's state dict into `to_save` in order to keep it immutable as Mapping requires...

Comment (sdesrozis) [reply ignite/contrib/handlers/lr_finder.py:341]: `lr_finder` just needs to restore state of trainer, that was a stupid remark !

**Pull request merged**